HPXMLtoOpenStudio/resources/vehicle.rb

```diff
@@ -0,0 +1,243 @@
+      vehicle.additional_properties.space = Geometry.get_space_from_location(vehicle.ev_charger.location, spaces)
+    end
+
+    # Get schedules to calculate effective discharge power
+    charging_schedule, discharging_schedule = get_ev_charging_schedules(runner, model, vehicle, schedules_file)
+    if charging_schedule.nil? && discharging_schedule.nil?
+      runner.registerWarning('Electric vehicle battery specified with no charging/discharging schedule provided; battery will not be modeled.')
+      return
+    end
+
+    # Calculate annual driving hours
+    if discharging_schedule.to_ScheduleFile.is_initialized
+      dis_sch = discharging_schedule.to_ScheduleFile.get
+      col = dis_sch.columnNumber - 1
+      discharge_vals = dis_sch.csvFile.get.getColumnAsStringVector(col).map(&:to_f)[1..]
+
+      # Scale based on timestep and schedule length
+      timestep_hours = dis_sch.minutesperItem.get.to_f / 60.0
+      driving_hrs = discharge_vals.sum * timestep_hours
+      schedule_hours = discharge_vals.size * timestep_hours
+      annual_driving_hours = driving_hrs / schedule_hours * UnitConversions.convert(1, 'yr', 'hr')
+    elsif discharging_schedule.to_ScheduleRuleset.is_initialized
+      model_year = model.yearDescription.get.assumedYear
+      year_start_date = OpenStudio::Date.new(OpenStudio::MonthOfYear.new('January'), 1, model_year)
+      year_end_date = OpenStudio::Date.new(OpenStudio::MonthOfYear.new('December'), 31, model_year)
+      discharge_day_schs = discharging_schedule.to_ScheduleRuleset.get.getDaySchedules(year_start_date, year_end_date)
+      annual_driving_hours = discharge_day_schs.sum { |day_sch| day_sch.timeSeries.values.sum }
+    end
+
+    # Check for discrepancy between inputted and calculated hours per week
+    sch_hours_per_week = annual_driving_hours / UnitConversions.convert(1, 'yr', 'day') * 7
+    hours_differ = sch_hours_per_week.round(1) != vehicle.hours_per_week.round(1)
+    if hours_differ
+      if !(vehicle.hours_per_week_isdefaulted || vehicle.ev_charging_weekday_fractions_isdefaulted || vehicle.ev_charging_weekend_fractions_isdefaulted)
+        runner.registerWarning("Electric vehicle hours per week inputted (#{vehicle.hours_per_week.round(1)}) do not match the hours per week calculated from the discharging schedule (#{sch_hours_per_week.round(1)}). The inputted hours per week value will be ignored.")
+      end
+      vehicle.hours_per_week = sch_hours_per_week
+    end
+
+    # Calculate effective discharge power and rated power output
+    # Scale the effective discharge power by 2.25 to assign the rated discharge power. This value reflects the maximum power adjustment allowed in the EMS EV discharge program at -17.8 C.
+    ev_annl_energy = vehicle.energy_efficiency * vehicle.miles_per_year # kWh/year
+    eff_discharge_power = UnitConversions.convert(ev_annl_energy / annual_driving_hours, 'kw', 'w') # W
+    vehicle.rated_power_output = eff_discharge_power * 2.25
```

Comment:
The rated power output is set to avoid ever limiting the actual discharge power at a timestep. This means that some combinations of inputs could result in very high and unrealistic effective discharge powers (eg, high miles, low hours == high annual energy over a short amount of hours). Granted, this would only happen under unrealistic combinations of inputs.